lib/TransferService.php

```diff
@@ -0,0 +1,114 @@
+use OCA\Guests\Db\Transfer;
+use OCA\Guests\Db\TransferMapper;
+use OCP\AppFramework\Db\DoesNotExistException;
+use OCP\AppFramework\QueryException;
+use OCP\BackgroundJob\IJobList;
+use OCP\IUser;
+use OCP\Notification\IManager as INotificationManager;
+use OCP\Share\IManager as IShareManager;
+use Psr\Container\ContainerInterface;
+use Psr\Log\LoggerInterface;
+
+class TransferService {
+	public function __construct(
+		private ContainerInterface $container,
+		private IShareManager $shareManager,
+		private INotificationManager $notificationManager,
+		private IJobList $jobList,
+		private TransferMapper $transferMapper,
+		private LoggerInterface $logger,
+	)
+	{
+	}
+
+	public function transfer(IUser $guestUser, IUser $user): void {
+		try {
+			/** @var OwnershipTransferService $ownershipTransferService */
+			$ownershipTransferService = $this->container->get(OwnershipTransferService::class);
+		} catch (QueryException $e) {
+			$this->logger->error('Could not resolve ownership transfer service to import guest user data', [
+				'exception' => $e,
+			]);
+			throw $e;
+		}
+
+		try {
+			$ownershipTransferService->transfer(
+				$guestUser,
+				$user,
+				'/',
+				null,
+				true,
+				true
+			);
+		} catch (TransferOwnershipException $e) {
+			$this->logger->error('Could not import guest user data', [
+				'exception' => $e,
+			]);
+			throw $e;
+		}
+
+		// Update incomming shares
+		$shares = $this->shareManager->getSharedWith($guestUser->getUID(), IShare::TYPE_USER);
+		foreach ($shares as $share) {
+			$share->setSharedWith($user->getUID());
+			$this->shareManager->updateShare($share);
+		}
+
+		$notification = $this->notificationManager->createNotification();
+		$notification
+			->setApp(Application::APP_ID)
+			->setSubject('data_migrated_to_system_user')
+			->setObject('user', $user->getEMailAddress())
+			->setDateTime(new \DateTime())
+			->setUser($user->getUID());
+		$this->notificationManager->notify($notification);
+	}
+
+	public function addTransferJob(IUser $author, IUser $source, string $target): void {
+		$transfer = new Transfer();
+		$transfer->setAuthor($author->getUID());
+		$transfer->setSource($source->getUID());
+		$transfer->setTarget($target);
+		$transfer->setStatus(Transfer::STATUS_WAITING);
+		/** @var Transfer $transfer */
+		$transfer = $this->transferMapper->insert($transfer);
+
+		$this->jobList->add(TransferJob::class, [
+			'id' => $transfer->getId(),
+		]);
+	}
+
+	public function getTransfer(IUser $source, string $target): ?Transfer {
+		try {
+			$transfer = $this->transferMapper->getBySource($source->getUID());
+			return $transfer;
+		} catch (DoesNotExistException $e) {
+			// Allow as this just means there is no pending transfer
+		}
+
+		try {
+			$transfer = $this->transferMapper->getByTarget($target);
+			return $transfer;
+		} catch (DoesNotExistException $e) {
+			// Allow as this just means there is no pending transfer
+		}
+
+		return null;
+	}
```

Comment:
Not sure that I get this method. Shouldn't the result be a transfer from source AND to target? In the current case, it seems that it could return a transfer either from source OR to target.

Reply:
I guess the reason is that, in `UserController`, want to ensure that no transfer exists for either users.
Maybe making it explicit in `UserController` would be better? By that, I mean copy-pasting the content of that method in `UserController::transfer`. It would be a bit longer, but easier to read.

Or maybe changing the method to: `getPendingTransfers(array $userList)`, and calling `getBySource` and `getByTarget` for each of them?

Reply:
Yes, the reason is to prevent any conflicts with the source or target users

Went with the simpler option 1